pkgs/build-support/node/fetch-pnpm-deps/default.nix

```diff
@@ -0,0 +1,115 @@
+  # to add more versions of this helper function in the future
+  pnpmVersions = [
+    (nodePackages.pnpm.overrideAttrs {
+      name = "pnpm-8.15.5";
+      src = fetchurl {
+        url = "https://registry.npmjs.org/pnpm/-/pnpm-8.15.5.tgz";
+        hash = "sha512-sFGjLH5pWDO4SSbTspuMylclS1ifBknYmcbp0O22cLkex+KkNFm65zdZu1zmGcMmbxFr+THOItHvF1mn5Fqpbw==";
+      };
+    })
+    (nodePackages.pnpm.overrideAttrs {
+      name = "pnpm-9.0.6";
+      src = fetchurl {
+        url = "https://registry.npmjs.org/pnpm/-/pnpm-9.0.6.tgz";
+        hash = "sha256-BiTjDv+GbN6zY7FQYb23/ZQlsXvBu0LCL19O/eoh9rM=";
+      };
+    })
+  ];
+
+  mkFetchPnpmDeps = pnpm:
+    {
+      inherit pnpm;
+
+      fetchPnpmDeps =
+        { src
+        , hash ? ""
+        , pname
+        , ...
+        } @ args:
+        let
+          args' = builtins.removeAttrs args [ "hash" "pname" ];
+          hash' =
+            if hash != "" then {
+              outputHash = hash;
+            } else {
+              outputHash = "";
+              outputHashAlgo = "sha256";
+            };
+        in
+        stdenvNoCC.mkDerivation (args' // {
+          name = "${pname}-pnpm-deps";
+
+          nativeBuildInputs = [
+            jq
+            moreutils
+            pnpm
+            cacert
+          ];
+
+          installPhase = ''
+            runHook preInstall
+
+            export HOME=$(mktemp -d)
+            pnpm config set store-dir $out
+            # Some packages produce platform dependent outputs. We do not want to cache those in the global store
+            pnpm config set side-effects-cache false
+            # As we pin pnpm versions, we don't really care about updates
+            pnpm config set update-notifier false
+            # pnpm is going to warn us about using --force
+            # --force allows us to fetch all dependencies including ones that aren't meant for our host platform
+            pnpm install --frozen-lockfile --ignore-script --force
+
+            runHook postInstall
+          '';
+
+          fixupPhase = ''
+            runHook preFixup
+
+            # Remove timestamp and sort the json files
+            rm -rf $out/v3/tmp
+            for f in $(find $out -name "*.json"); do
+              sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+              jq --sort-keys . $f | sponge $f
+            done
+
+            runHook postFixup
+          '';
+
+          passthru = {
+            inherit pnpm;
+          };
+
+          dontConfigure = true;
+          dontBuild = true;
+          outputHashMode = "recursive";
+        } // hash');
+
+      pnpmConfigHook = makeSetupHook
+        {
+          name = "pnpm-config-hook";
+          propagatedBuildInputs = [ pnpm ];
+        } ./pnpm-config-hook.sh;
+
+      __functor = self: self.fetchPnpmDeps;
+    };
+
+  mkFetchPnpmDepsAttrs = fnVersion: pnpm: nameValuePair "fetchPnpmDeps${toString fnVersion}" (mkFetchPnpmDeps pnpm);
```

Comment:
i would strongly suggest going back to the original naming scheme of lockfile version -- or alternative, use the pnpm version
using the really arbitrary index of a version in a list gives callers no real information on what that variant is for